File: prefy/src/com/hendraanggrian/prefy/Prefy.java
package com.hendraanggrian.prefy;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.Map;
import java.util.WeakHashMap;

/**
 * Create an instance of {@link WritablePreferences} or {@link EditablePreferences}
 * and bind its values to JVM fields with this object.
 */
public final class Prefy {
    private static Map<Class<?>, Constructor<PreferencesSaver>> BINDINGS;
    private static PreferencesLogger LOGGER;

    private Prefy() {
    }

    /**
     * Set logging behavior using {@link PreferencesLogger} instance.
     *
     * @param logger custom or pre-determined logger (e.g: {@link PreferencesLogger#SYSTEM}).
     */
    public static void setLogger(PreferencesLogger logger) {
        LOGGER = logger;
    }

    static void info(String format, Object... args) {
        if (LOGGER != null) {
            LOGGER.info(String.format(format, args));
        }
    }

    static void warn(String format, Object... args) {
        if (LOGGER != null) {
            LOGGER.info(String.format(format, args));
        }
    }

    /**
     * Bind fields annotated with {@link BindPreference} with values from this preferences.
     *
     * @param target fields' owner.
     * @return saver instance to apply changes made to the fields.
     * @throws RuntimeException when constructor of binding class cannot be found.
     * @receiver source preferences.
     */
    public static PreferencesSaver bind(ReadablePreferences source, Object target) {
        final Class<?> targetCls = target.getClass();
        final String targetClsName = targetCls.getName();
        info("Looking up binding for %s", targetClsName);
        final Constructor<PreferencesSaver> constructor = findBindingConstructor(targetCls);
        if (constructor == null) {
            info("%s binding not found, returning empty saver.", targetClsName);
            return PreferencesSaver.EMPTY;
        }
        try {
            return constructor.newInstance(source, target);
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Unable to invoke " + constructor, e);
        } catch (InstantiationException e) {
            throw new RuntimeException("Unable to invoke " + constructor, e);
        } catch (InvocationTargetException e) {
            final Throwable cause = e.getCause();
            if (cause instanceof RuntimeException)
                throw (RuntimeException) cause;
            if (cause instanceof Error)
                throw (Error) cause;
            throw new RuntimeException("Unable to create constructor instance.", cause);
        }
    }

    @SuppressWarnings("unchecked")
    static Constructor<PreferencesSaver> findBindingConstructor(Class<?> cls) {
        if (BINDINGS == null) BINDINGS = new WeakHashMap<>();
        Constructor<PreferencesSaver> binding = BINDINGS.get(cls);
        if (binding != null) {
            info("HIT: Cache found in binding weak map.");
            return binding;
        }
        final String clsName = cls.getName();
        if (clsName.startsWith("android.") || clsName.startsWith("java.")) {
            info("MISS: Reached framework class. Abandoning search.");
            return null;
        }
        try {
            binding = (Constructor<PreferencesSaver>) cls.getClassLoader()
                .loadClass(clsName + "_PreferencesBinding")
                .getConstructor(ReadablePreferences.class, cls);
            info("HIT: Loaded binding class, caching in weak map.");
        } catch (ClassNotFoundException e) {
            final Class<?> supercls = cls.getSuperclass();
            warn("Not found. Trying superclass %s ...", supercls.getName());
            binding = findBindingConstructor(supercls);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("Unable to find binding constructor for " + clsName, e);
        }
        if (binding == null) {
            throw new IllegalStateException("Unable to find preferences binding, " +
                "is `prefy-compiler` correctly installed?");
        }
        BINDINGS.put(cls, binding);
        return binding;
    }
}

File: prefy/src/com/hendraanggrian/prefy/Prefy.kt
package com.hendraanggrian.prefy

import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.util.WeakHashMap

/** Create an instance of [WritablePreferences] or [EditablePreferences] and bind its values to JVM fields with this object. */
object Prefy {
    private var BINDINGS: MutableMap<Class<*>, Constructor<PreferencesSaver>>? = null
    private var LOGGER: PreferencesLogger? = null

    /**
     * Set logging behavior using [PreferencesLogger] instance.
     * @param logger custom or pre-determined logger (e.g: [PreferencesLogger.System]).
     */
    fun setLogger(logger: PreferencesLogger) {
        LOGGER = logger
    }

    internal fun info(message: String) {
        LOGGER?.info(message)
    }

    internal fun warn(message: String) {
        LOGGER?.warn(message)
    }

    /**
     * Bind fields annotated with [BindPreference] with values from this preferences.
     * @param source preferences to extract.
     * @param target fields' owner.
     * @return saver instance to apply changes made to the fields.
     * @throws RuntimeException when constructor of binding class cannot be found.
     */
    fun bind(source: ReadablePreferences, target: Any): PreferencesSaver {
        val cls = target.javaClass
        val clsName = cls.simpleName
        info("Looking up binding for $clsName")
        val constructor = findBindingConstructor(cls)
        if (constructor == null) {
            info("$clsName binding not found, returning empty saver.")
            return PreferencesSaver.EMPTY
        }
        try {
            return constructor.newInstance(source, target)
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Unable to invoke $constructor", e)
        } catch (e: InstantiationException) {
            throw RuntimeException("Unable to invoke $constructor", e)
        } catch (e: InvocationTargetException) {
            when (val cause = e.cause) {
                is RuntimeException -> throw cause
                is Error -> throw cause
                else -> throw RuntimeException("Unable to create binding instance.", cause)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun findBindingConstructor(cls: Class<*>): Constructor<PreferencesSaver>? {
        if (BINDINGS == null) {
            BINDINGS = WeakHashMap()
        }
        var binding = BINDINGS!![cls]
        if (binding != null) {
            info("HIT: Cache found in binding weak map.")
            return binding
        }
        val clsName = cls.simpleName
        if (clsName.startsWith("android.") || clsName.startsWith("java.")) {
            info("MISS: Reached framework class. Abandoning search.")
            return null
        }
        try {
            binding = cls.classLoader!!
                .loadClass(clsName + BindPreference.SUFFIX)
                .getConstructor(ReadablePreferences::class.java, cls) as Constructor<PreferencesSaver>
            info("HIT: Loaded binding class, caching in weak map.")
        } catch (e: ClassNotFoundException) {
            val supercls = cls.superclass!!
            warn("Not found. Trying superclass ${supercls.simpleName} ...")
            binding = findBindingConstructor(supercls)
        } catch (e: NoSuchMethodException) {
            throw RuntimeException("Unable to find binding constructor for $clsName", e)
        }
        BINDINGS!![cls] = checkNotNull(binding) {
            "Unable to find preferences binding, is `prefy-compiler` correctly installed?"
        }
        return binding
    }
}
